dig.go

```diff
@@ -332,15 +341,25 @@ func setRand(r *rand.Rand) Option {
 }
 
 func (c *Container) knownTypes() []reflect.Type {
-	typeSet := make(map[reflect.Type]struct{}, len(c.providers))
-	for k := range c.providers {
-		typeSet[k.t] = struct{}{}
+	getKnowTypes := func(c *Container) []reflect.Type {
+		typeSet := make(map[reflect.Type]struct{}, len(c.providers))
+		for k := range c.providers {
+			typeSet[k.t] = struct{}{}
+		}
+
+		types := make([]reflect.Type, 0, len(typeSet))
+		for t := range typeSet {
+			types = append(types, t)
+		}
+
+		return types
 	}
 
-	types := make([]reflect.Type, 0, len(typeSet))
-	for t := range typeSet {
-		types = append(types, t)
+	types := make([]reflect.Type, 0)
+	for _, c := range append(c.children, c) {
+		types = append(types, getKnowTypes(c)...)
```

Comment:
Not a major issue but I don't think this level of change was necessary. You
can let recursion handle the work here like with the other cases.

```
func (c *Container) knownTypes() []reflect.Type {
  [previous code]

  for _, c := range c.children {
    types = append(types, c.knownTypes()...)
  }

  sort.Sort(byTypeName(types))
  return types
}
```

The result is more readable in exchange for wasted effort in sorting the
children's slices, but that should be fine since this is only used on the
error path.